### PR TITLE
CTCP-4244: Add abstraction to allow encrypted components to be hidden from presentation view

### PR DESCRIPTION
**Pull request by drnaylor-oc**
This PR primarily adds field level encryption to the `messages[*].body` field in a way that will also read unencrypted strings, allowing for Trader Test to continue without wiping data, but with a switch so that this can never happen on Production.

This PR also splits out Mongo models from presentation models so that the markers for encryption can be managed separately, and inserts a service layer between the repo and the controllers to allow for this conversion.

This also allows for the `PresentationFormats` class to be deleted, as we don't share top level models between Mongo and presentation any more (it's easier to see the separation when the models are separate, rather than just the formats).

The `messageSender` is no longer important, so is not indexed. The LRN remains indexed, as we are able to filter on that.

Of note in the config:

* `encryption.enabled` will be `true` everywhere. If it is turned off, a warning will be logged, which I intend to add to the PD alerts to at least make it obvious when someone does something daft.
* `encryption.tolerant-read` will be `false` on production.
* `encryption.key` will be set separately on all environments, I'll generate a 256 bit key for each environment.

**Please make sure you check and test this thoroughly before approving. This is an invasive, potentially data destroying change.** I have tested against the regression tests for both small and large messages and had no errors, but don't take that for granted.

### CODE REVIEW

**Review by derekspalmer — APPROVED** [review]
I've had a good look through  ... and and am still looking but seems ok.
I'm still trying to test it but will approve on basis needed today.

**Comment by drnaylor-oc** [comment]
Not needed today, I'm content to wait.

**Review by derekspalmer — COMMENTED** [review]
There's a lot of change here: I would probably have kept it simple initially and just added the encryption to ensure it's working, and then, when we have a high level of confidence in it, add the refactoring/tech debt stuff.  Having said that, having looked at the crypto docs in Confluence, the crypto service itself, and other services that use it,  it looks fine.
On a side-point, I would only encrypt data in Production and Staging. If a trader posts sensitive info to External Test, that's their problem, as long as we've told them not to. We need External Test to be as 'friendly' as possible for the team (and those who come after us) to identify issues quickly.

File: app/uk/gov/hmrc/transitmovements/config/AppConfig.scala
Comment (drnaylor-oc) [reply app/uk/gov/hmrc/transitmovements/config/AppConfig.scala:33]: They're used by bootstrap components, so we can't. We just don't use them in our components.

**Comment by drnaylor-oc** [comment]
> There's a lot of change here: I would probably have kept it simple initially and just added the encryption to ensure it's working, and then, when we have a high level of confidence in it, add the refactoring/tech debt stuff

I disagree in this instance. The ultimate problem is that the presentation layer models and the mongo models probably shouldn't have been the same model (and I'm just as culpable in that), and actually, all this refactoring does is add a layer in the middle to do the model translation because the models have indeed diverged.

My original thought was to make the `Reads` and `Writes` diverge further, but that actually ended up being a really messy hack. I felt it was better to do things properly in this instance.

> If a trader posts sensitive info to External Test, that's their problem, as long as we've told them not to. We need External Test to be as 'friendly' as possible for the team (and those who come after us) to identify issues quickly.

Unfortunately, that's not how Security see it. ET is a live environment and there is the potential for third parties to add PII - indeed, PII that might not be theirs. In fact, the reality is that they have indeed used previous movements as test beds. 

I get it though. I'll have to talk to various people to see what they think, but I wouldn't be hopeful about it.

For Development and QA though, I can agree with that, and for that reason until we decide what we're doing about it, I'll hold off the merge button.

**Pull request merged**
